File: algorithms/B/Battleships in a Board/Battleships in a Board.cpp
class Solution {
    int m, n;
    vector<int> dx{-1, 0, 1, 0};
    vector<int> dy{0, 1, 0, -1};
public:
    int countBattleships(vector<vector<char>>& board) {
        
        int cnt = 0;
        m = board.size(), n = board[0].size();
        
        for(int i = 0; i < m; i++)
            for(int j = 0; j < n; j++)
                if(board[i][j] == 'X')
                    cnt++, helper(board, i, j);
        
        return cnt;
    }
    
    void helper(vector<vector<char>> &board, int i, int j)
    {
        board[i][j] = '.';
        for(int k = 0; k < dx.size(); k++)
        {
            int x = i + dx[k], y = j + dy[k];
            if(x >= 0 && y >= 0 && x < m && y < n && board[x][y] == 'X')
                helper(board, x, y);
        }
    }
};

File: scripts/algorithms/B/Battleships in a Board/Battleships in a Board.java
// Runtime: 4 ms (Top 10.56%) | Memory: 44.7 MB (Top 26.11%)
class Solution {
    public int countBattleships(char[][] board) {
            int result = 0;
            for(int i = 0;i<board.length;i++)
            {
                for(int j = 0;j<board[i].length;j++)
                {
                    if(board[i][j] == 'X')
                    {
                        remover(i, j, board);
                        result++;
                    }

                }
            }
            return result;

        }
        public void remover(int i , int j, char[][] board)
        {
            if(i<0 || i>= board.length || j<0 || j>=board[0].length || board[i][j] == '.')
            {
                return;
            }
            board[i][j] = '.';
            remover(i+1, j, board);
            remover(i, j+1, board);
        }
}